### PR TITLE
Automatically download ollama if not installed

### PR DESCRIPTION
**Pull request by rockwotj**
Stacked on #2752

To improve the 0 to 60second experience we want to auto download ollama for devs if it's not installed locally.

### CODE REVIEW

**Comment by emaxerrno** [comment]
SUPER SUPER COOL.

File: internal/impl/ollama/base_processor.go [hunk 1]
```diff
@@ -86,6 +116,39 @@ var ollamaProcess = singleton.New(singleton.Config[*exec.Cmd]{
 		return cmd.Wait()
 	}})
 
+func downloadOllama(ctx context.Context, path string) error {
+	var url string
+	const baseURL string = "https://github.com/ollama/ollama/releases/download/v0.3.1/ollama"
+	switch runtime.GOOS {
```
Comment (emaxerrno) [inline internal/impl/ollama/base_processor.go:122]: should this be configurable.

Comment (emaxerrno) [reply internal/impl/ollama/base_processor.go:122]: with a default.

Comment (rockwotj) [reply internal/impl/ollama/base_processor.go:122]: Good point. This will be the default, and I'll add a string to support overriding.

**Review by mihaitodor — COMMENTED** [review]
Thanks Tyler! Just left one comment on how to do file stuff

File: internal/impl/ollama/base_processor.go [hunk 2]
```diff
@@ -86,6 +116,39 @@ var ollamaProcess = singleton.New(singleton.Config[*exec.Cmd]{
...
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to download ollama binary: %w", err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 {
+		return fmt.Errorf("failed to download ollama binary: status_code=%d", resp.StatusCode)
+	}
+	ollama, err := os.OpenFile(path, os.O_CREATE|os.O_TRUNC|os.O_RDWR, 0777)
```
[the remark this answers is not in the document]
Comment (rockwotj) [reply internal/impl/ollama/base_processor.go:143]: the `service.FS` interface doesn't allow for writing to the filesystem. How do I work around that? Do I need to add that to Benthos core?

[the remark this answers is not in the document]
Comment (rockwotj) [reply internal/impl/ollama/base_processor.go]: I just am casting to `io.Writer` for now

File: internal/impl/ollama/base_processor.go (hunk 1 above)
Comment (rockwotj) [reply internal/impl/ollama/base_processor.go]: Done

File: internal/impl/ollama/base_processor.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mihaitodor) [reply internal/impl/ollama/base_processor.go]: Oh, good point... I'm actually not sure if there's any plan to expand that interface. Will defer to @Jeffail on this stuff.

File: internal/impl/ollama/base_processor.go [hunk 3]
```diff
@@ -55,22 +65,109 @@ func (co *commandOutput) Write(b []byte) (int, error) {
...
+		p, err := c.fs.Stat(path)
+		if err != nil || p.IsDir() {
+			continue
+		}
+		// Check that the file is executable
+		if err := syscall.Access(path, 0x1); err == nil {
+			return path, nil
+		}
+	}
+	return "", exec.ErrNotFound
```
Comment (mihaitodor) [inline internal/impl/ollama/base_processor.go:98]: If possible, can we have an explanation here on why we have to traverse `$PATH` manually and check if the file is executable? I just tested https://pkg.go.dev/os/exec#LookPath and I'm not sure how that function differs.

File: internal/impl/ollama/base_processor.go [hunk 4]
```diff
@@ -55,22 +65,109 @@ func (co *commandOutput) Write(b []byte) (int, error) {
...
+}
+
+func (c *runOllamaConfig) downloadOllama(ctx context.Context, path string) error {
+	var url string
+	const baseURL string = "https://github.com/ollama/ollama/releases/download/v0.3.1/ollama"
```
Comment (mihaitodor) [inline internal/impl/ollama/base_processor.go:103]: Should we consider exposing the version as a config parameter?

**Review by mihaitodor — APPROVED** [review]
Thanks Tyler! I left a few small comments, but it should be good to go otherwise.

File: internal/impl/ollama/base_processor.go (hunk 3 above)
Comment (rockwotj) [reply internal/impl/ollama/base_processor.go:98]: Yeah, the difference is that we're using our filesystem abstractions - will add a comment

File: internal/impl/ollama/base_processor.go (hunk 4 above)
Comment (rockwotj) [reply internal/impl/ollama/base_processor.go:103]: Oh probably since this is already old 😄

Comment (rockwotj) [reply internal/impl/ollama/base_processor.go:103]: Well you can already provide a override URL to download from whereever, so I think that is a good enough workaround for now

File: internal/impl/ollama/base_processor.go (hunk 3 above)
Comment (mihaitodor) [reply internal/impl/ollama/base_processor.go:100]: Might be a good idea to have it in Benthos. Smth to check with @Jeffail for a future refactor

File: internal/impl/ollama/base_processor.go [hunk 5]
```diff
@@ -55,22 +65,111 @@ func (co *commandOutput) Write(b []byte) (int, error) {
...
+	switch runtime.GOOS {
+	case "darwin":
+		// They ship an universal executable for darwin
+		url = baseURL + "-darwin"
+	case "linux":
+		url = fmt.Sprintf("%s-%s-%s", baseURL, runtime.GOOS, runtime.GOARCH)
+	default:
+		return fmt.Errorf("automatic download of ollama is not supported on %s, please download ollama manually", runtime.GOOS)
+	}
+	if c.downloadURL != "" {
```
Comment (mihaitodor) [inline internal/impl/ollama/base_processor.go:115]: Should we maybe move this check above the switch just in case people have a download link for Windows?

Comment (rockwotj) [reply internal/impl/ollama/base_processor.go]: Oh good catch! Fixed.

**Review by mihaitodor — APPROVED** [review]
LGTM :shipit:

**Pull request merged**